File: day13/part2.rb
if ARGV.empty?
  puts "usage: ruby #{__FILE__} FILENAME"
  exit 1
end

prog_l = File.read(ARGV[0]).strip.split(',').map(&:to_i)
prog = Hash.new { |h, k| h[k] = 0 }
prog_l.each_with_index do |v, i|
  prog[i] = v
end
rb = pc = 0

def param(prog, pc, rb, instr, n)
  mode = instr / (10**(n + 1)) % 10
  case mode
  when 0
    prog[prog[pc + n]]
  when 1
    prog[pc + n]
  when 2
    prog[rb + prog[pc + n]]
  else
    raise StandardError, 'unreachable!'
  end
end

def store(prog, pc, rb, instr, n)
  mode = instr / (10**(n + 1)) % 10
  case mode
  when 0
    prog[pc + n]
  when 2
    rb + prog[pc + n]
  else
    raise StandardError, 'unreachable!'
  end
end

TILES = { 0 => ' ', 1 => '█', 2 => '▦', 3 => '▂', 4 => '○' }.freeze
board = Hash.new { |h, k| h[k] = 0 }
max_x = 0
max_y = 0
read_queue = []
score = 0
ball_pos = 0
paddle_pos = 0

prog[0] = 2

loop do
  instr = prog[pc]
  opc = instr % 100
  case opc
  when 99
    puts score
    exit 0
  when 1
    prog[store(prog, pc, rb, instr, 3)] = (
        param(prog, pc, rb, instr, 1) + param(prog, pc, rb, instr, 2)
    )
    pc += 4
  when 2
    prog[store(prog, pc, rb, instr, 3)] = (
        param(prog, pc, rb, instr, 1) * param(prog, pc, rb, instr, 2)
    )
    pc += 4
  when 3
    if ENV['DRAW']
      (0..max_y).each do |y|
        (0..max_x).each do |x|
          print(TILES[board[[x, y]]])
        end
        puts
      end
    end
    val = if ball_pos > paddle_pos
            1
          elsif ball_pos == paddle_pos
            0
          else
            -1
          end
    prog[store(prog, pc, rb, instr, 1)] = val
    pc += 2
  when 4
    read_queue.push(param(prog, pc, rb, instr, 1))
    if read_queue.length == 3
      x, y, val = read_queue
      if x == -1 && y.zero?
        score = val
      else
        board[[x, y]] = val
        if val == 4
          ball_pos = x
        elsif val == 3
          paddle_pos = x
        end
      end
      max_x = [max_x, x].max
      max_y = [max_y, y].max
      read_queue.clear
    end
    pc += 2
  when 5
    if param(prog, pc, rb, instr, 1) != 0
      pc = param(prog, pc, rb, instr, 2)
    else
      pc += 3
    end
  when 6
    if param(prog, pc, rb, instr, 1).zero?
      pc = param(prog, pc, rb, instr, 2)
    else
      pc += 3
    end
  when 7
    if param(prog, pc, rb, instr, 1) < param(prog, pc, rb, instr, 2)
      prog[store(prog, pc, rb, instr, 3)] = 1
    else
      prog[store(prog, pc, rb, instr, 3)] = 0
    end
    pc += 4
  when 8
    if param(prog, pc, rb, instr, 1) == param(prog, pc, rb, instr, 2)
      prog[store(prog, pc, rb, instr, 3)] = 1
    else
      prog[store(prog, pc, rb, instr, 3)] = 0
    end
    pc += 4
  when 9
    rb += param(prog, pc, rb, instr, 1)
    pc += 2
  else
    raise StandardError, 'unreachable!'
  end
end

File: day13/part2.py
import argparse
import collections
import curses
import hashlib
import os
import time
from typing import Callable
from typing import Dict
from typing import List
from typing import Tuple
from typing import Union

from support import timing


def run(
        prog: Dict[int, int],
        in_fn: Callable[[], int],
        out_fn: Callable[[int], None],
) -> None:
    def parameter(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[prog[pc + n]]
        elif mode == 1:
            return prog[pc + n]
        elif mode == 2:
            return prog[rb + prog[pc + n]]
        else:
            raise NotImplementedError(mode)

    def store(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[pc + n]
        elif mode == 2:
            return rb + prog[pc + n]
        else:
            raise NotImplementedError(mode)

    rb = 0
    pc = 0
    while pc < len(prog):
        instr = prog[pc]
        opc = instr % 100
        if opc == 99:
            return
        elif opc == 1:
            prog[store(instr, 3)] = parameter(instr, 1) + parameter(instr, 2)
            pc += 4
        elif opc == 2:
            prog[store(instr, 3)] = parameter(instr, 1) * parameter(instr, 2)
            pc += 4
        elif opc == 3:
            prog[store(instr, 1)] = in_fn()
            pc += 2
        elif opc == 4:
            out_fn(parameter(instr, 1))
            pc += 2
        elif opc == 5:
            if parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 6:
            if not parameter(instr, 1):
                pc = parameter(instr, 2)
            else:
                pc += 3
        elif opc == 7:
            if parameter(instr, 1) < parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 8:
            if parameter(instr, 1) == parameter(instr, 2):
                prog[store(instr, 3)] = 1
            else:
                prog[store(instr, 3)] = 0
            pc += 4
        elif opc == 9:
            rb += parameter(instr, 1)
            pc += 2
        else:
            raise AssertionError(f'unreachable? {prog} {pc}')
    raise AssertionError(f'unreachable? {prog} {pc}')


KEYS: Dict[Union[int, str], int] = {curses.KEY_LEFT: -1, curses.KEY_RIGHT: 1}
TILES = {0: ' ', 1: '█', 2: '▦', 3: '▂', 4: '○'}


def c_main(stdscr: 'curses._CursesWindow', prog_d: Dict[int, int]) -> int:
    board: Dict[Tuple[int, int], int] = collections.defaultdict(int)
    read_queue: List[int] = []
    score = 0
    ball_pos = 0
    paddle_pos = 0

    def draw_score(val: int) -> None:
        nonlocal score
        score = val
        stdscr.addstr(curses.LINES - 1, 0, f'score: {score}')

    def in_fn() -> int:
        if os.environ.get('MODE') == 'manual':
            return KEYS.get(stdscr.get_wch(), 0)
        elif os.environ.get('MODE') == 'fast':
            stdscr.refresh()
            time.sleep(.001)
        elif os.environ.get('MODE') == 'normal':
            stdscr.refresh()
            time.sleep(.05)
        if ball_pos > paddle_pos:
            return 1
        elif ball_pos == paddle_pos:
            return 0
        else:
            return -1

    def out_fn(n: int) -> None:
        nonlocal ball_pos
        nonlocal paddle_pos

        read_queue.append(n)
        if len(read_queue) == 3:
            x, y, val = read_queue
            if (x, y) == (-1, 0):
                draw_score(val)
            else:
                board[x, y] = val
                if val == 2:
                    pair = hashlib.sha256(f'{x}{y}'.encode()).digest()[0] % 16
                    color = curses.color_pair(pair)
                else:
                    color = 0

                if val == 4:
                    ball_pos = x
                elif val == 3:
                    paddle_pos = x
                stdscr.addstr(y, x, TILES[val], color)
            read_queue.clear()

    prog_d[0] = 2
    draw_score(0)
    curses.curs_set(0)
    curses.use_default_colors()
    for color in range(1, 16):
        curses.init_pair(color, color, -1)
    run(prog_d, in_fn, out_fn)
    return score


def compute(s: str) -> int:
    prog = [int(part) for part in s.strip().split(',')]
    prog_d = collections.defaultdict(int, enumerate(prog))
    return curses.wrapper(c_main, prog_d)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file')
    args = parser.parse_args()

    with open(args.data_file) as f, timing():
        print(compute(f.read()))

    return 0


if __name__ == '__main__':
    exit(main())
